### ISSUE TITLE
retain_grad does not work when the starting node of the computational graph is numpy array 

### ISSUE THREAD

**Issue by fiarabbit** [chainer/chainer#3006]
I found that  I cannot access the grad of an intermediate node though `retain_grad=True` is specified when the starting node of the computational graph is numpy.ndarray.

It has to be noted that when the starting node is `chainer.variable.Variable`, the intermediate gradient can be accessed.

* Conditions
  - Chainer version
chainer v2.0.1
  - OS/Platform
Ubuntu 16.04 LTS

* Code to reproduce
``` test.py
from chainer import functions as F
import numpy
from chainer import Variable

# OK
n0_Variable = Variable(numpy.array([1, 2, 3], 'f'))
n1_Variable = F.tanh(n0_Variable)
n2_Variable = F.sum(n1_Variable)
n2_Variable.backward(True)

# NG
n0_numpy = numpy.array([1, 2, 3], 'f')
n1_numpy = F.tanh(n0_numpy)
n2_numpy = F.sum(n1_numpy)
n2_numpy.backward(True)

print("n1_Variable:\n" + n1_Variable.debug_print() + "\n")
print("n1_numpy:\n" + n1_numpy.debug_print() + "\n")
```
* Error messages, stack traces, or logs
``` output.py
n1_Variable:
<variable at 0x7fb8c0c014a8>
- device: CPU
- backend: <class 'numpy.ndarray'>
- shape: (3,)
- dtype: float32
- statistics: mean=0.90689212, std=0.10351910
- grad: mean=1.00000000, std=0.00000000

n1_numpy:
<variable at 0x7fb8c0c01278>
- device: CPU
- backend: <class 'numpy.ndarray'>
- shape: (3,)
- dtype: float32
- statistics: mean=0.90689212, std=0.10351910
- grad: None
```

**Comment by delta2323** [comment]
Thank you for your report. I could reproduce the result.

**Comment by delta2323** [comment]
Chainer `Variable` has an attribute `requires_grad` that indicates if `grad` is set in backward computation. Note that `requires_grad` precedes `retain_grad` of `Variable.backward` (i.e. `grad` will not be set when `requires_grad` is `False` even if `retain_grad` is `True`).

If we put bare NumPy ndarray's to a Chainer `Function`, it is automatically wrapped with a `Variable` with `requires_grad` `False`. Also, if `requires_grad` of all input variables are `False`, the output `variable` will also set `False` to `requires_grad`. 

Actually, if we set `requires_grad=True` to `n0_Variable`, the output will be the same as numpy-input case.

**Comment by fiarabbit** [comment]
I understand your comment. I guess `retain_grad=True` in your comment may be  `retain_grad=False`. The following is the reproduction.

```requires_grad.py
from chainer import functions as F
import numpy
from chainer import Variable

# requires_grad=True is the same with default Variable
n0_True = Variable(numpy.array([1, 2, 3], 'f'), requires_grad=True)
n1_True = F.tanh(n0_True)
n2_True = F.sum(n1_True)
n2_True.backward(True)

print("n1_True:\n" + n1_True.debug_print() + "\n")

# requires_grad=False is the same with default numpy.ndarray
n0_False = Variable(numpy.array([1, 2, 3], 'f'), requires_grad=False)
n1_False = F.tanh(n0_False)
n2_False = F.sum(n1_False)
n2_False.backward(True)

print("n1_False:\n" + n1_False.debug_print() + "\n")
```

But why the default behavior of `numpy.ndarray` and `Variable` is different?
And how do I change the property of `requires_grad` when I use `numpy.ndarray` as the starting node of computational graph?

**Comment by delta2323** [comment]
This specification expect that if users put NumPy as a variable node of the computational graph, users do not want its gradient. But personally I come to think it oversees the use case in which users want to get the gradient of intermediate variables as you did.

Chainer forbids to set manually `requires_grad` other than `Variable.__init__`. Suppose the computational graph is a linear graph, if we set `requires_grad=False` to one of intermediate variables, Chainer does not set grads to upstream variables of the variable.

**Comment by fiarabbit** [comment]
Thank you for your kind explanation!
